snapcraft/plugins/v2/qmake.py

```diff
@@ -0,0 +1,77 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2020 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The qmake plugin is useful for building qmake-based parts.
+
+These are projects that are built using .pro files.
+
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - qmake-parameters:
+      (list of strings)
+      additional options to pass to the qmake invocation.
+"""
+
+from typing import Any, Dict, List, Set
+
+from snapcraft.plugins.v2 import PluginV2
+
+
+class QMakePlugin(PluginV2):
+    @classmethod
+    def get_schema(cls) -> Dict[str, Any]:
+        return {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "qmake-parameters": {
+                    "type": "array",
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                }
+            },
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {"g++", "make", "qt5-qmake", "qtbase5-dev"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {"QT_SELECT": "qt5"}
+
+    def _get_qmake_configure_command(self) -> str:
+        cmd = [
+            "qmake",
+            'QMAKE_CFLAGS+="$CFLAGS"',
+            'QMAKE_CXXFLAGS+="$CXXFLAGS"',
+            'QMAKE_LFLAGS+="$LDFLAGS"',
+        ] + self.options.qmake_parameters
+        return " ".join(cmd)
+
+    def get_build_commands(self) -> List[str]:
+        return [
+            self._get_qmake_configure_command(),
+            '(unset CFLAGS CXXFLAGS LDFLAGS; make -j"${SNAPCRAFT_PARALLEL_BUILD_COUNT}")',
```

Comment:
I guess unset is done as it is already done in _get_qmake_conflgure_commands. Mind adding a comment for that if it is the case, or if there is another reason, why the unset is done for future observers?

Reply:
I expect that the exported ones may conflict with qmake generated ones... but certainly would appreciate a comment :D

This is certainly a fine approach.  FWIW when I need to do something like this, i'll use `env -u <var1>...`

Reply:
In particular, the generated Makefile generated by qmake for the sample snap contains:

```
CFLAGS        = -pipe -O2 -fPIC $(DEFINES)
CXXFLAGS      = -pipe -O2 -fPIC $(DEFINES)
```

Having the environment variables set during the build will override the values from the Makefile, potentially miscompiling the project.

I would not be surprised if the same is true for a few other plugins.